### ISSUE TITLE
AddTenantTheme not working with REST API in SP2019 OnPrem

### ISSUE THREAD

**Issue by UrsWedershoven** [SharePoint/sp-dev-docs#4624]
#### Category
- [x] Question
- [ ] Typo
- [x] Bug
- [ ] Additional article idea

#### Expected or Desired Behavior

I want to add a modern theme to a Communications site in SharePoint Server 2019 as there is no explicit limitation given in the documentation that says this is only working for SharePoint Online. The REST endpoint is available.

#### Observed Behavior
When using the REST API together with the example given [here](https://docs.microsoft.com/de-de/sharepoint/dev/declarative-customization/site-theming/sharepoint-site-theming-rest-api) I only get 
{"@odata.context":"urlgoeshere/sites/siteurl/_api/$metadata#Edm.Boolean","value":false}
and nothing happens.

There are no error messages or any further information on anything missing or so.

Is there any way to use this function at all in an onPremises environment? Any alternatives?

**Comment by andrewconnell** [comment]
@JeremyKelley Are you aware if this API is only on SPO, or if it also exists on SP 2019? The doc the OP referenced seems to imply only SPO, but it doesn't explicitly state where this API exists.

**Comment by UrsWedershoven** [comment]
Any news on this one?
The endpoints listed in the documentation are generally available in my onPremises farm. I can find them all in the API metadata.

**Comment by tigertop** [comment]
Hi,
I was struggling with this API for a few days now. Found a note [here](https://docs.microsoft.com/en-us/sharepoint/dev/declarative-customization/site-design-overview) that states "Site designs and site scripts are currently only supported by SharePoint Online.". I am running SharePoint with April 2020 update and I have the same behavior as you had.

The positive side is that [AppTenantTheme](https://docs.microsoft.com/en-us/sharepoint/dev/declarative-customization/site-theming/sharepoint-site-theming-rest-api#addtenanttheme) api endpoint is working, and I am able ti apply my custom made theme using [this](https://developer.microsoft.com/en-us/fluentui#/styles/themegenerator?fabricVer=5) tool.

There are two nuances with that API. First you'll need to call the "<siteURL>/_api/contextinfo" and to get the FormDigest and second you need to stringify the JSON ([example tool](https://onlinetexttools.com/json-stringify-text)) and POST it inside this payload:
`
{
	"name":"feeee",
	"themeJson":"{\"palette\" : {\"themePrimary\": \"#e30613\",\"themeLighterAlt\": \"#fef4f5\",\"themeLighter\": \"#fbd4d6\",\"themeLight\": \"#f7afb4\",\"themeTertiary\": \"#ee646d\",\"themeSecondary\": \"#e6222f\",\"themeDarkAlt\": \"#cc0613\",\"themeDark\": \"#ac0510\",\"themeDarker\": \"#7f040c\",\"neutralLighterAlt\": \"#f8f8f8\",\"neutralLighter\": \"#f4f4f4\",\"neutralLight\": \"#eaeaea\",\"neutralQuaternaryAlt\": \"#dadada\",\"neutralQuaternary\": \"#d0d0d0\",\"neutralTertiaryAlt\": \"#c8c8c8\",\"neutralTertiary\": \"#dfb0f8\",\"neutralSecondary\": \"#c065f1\",\"neutralPrimaryAlt\": \"#a522eb\",\"neutralPrimary\": \"#9a07e8\",\"neutralDark\": \"#7405b0\",\"black\": \"#560482\",\"white\": \"#ffffff\",\"primaryBackground\": \"#ffffff\",\"primaryText\": \"#9a07e8\",\"bodyBackground\": \"#ffffff\",\"bodyText\": \"#9a07e8\",\"disabledBackground\": \"#f4f4f4\",\"disabledText\": \"#c8c8c8\"}}"
}
`

**Comment by kitzler-walli** [comment]
> Hi,
> I was struggling with this API for a few days now. Found a note [here](https://docs.microsoft.com/en-us/sharepoint/dev/declarative-customization/site-design-overview) that states "Site designs and site scripts are currently only supported by SharePoint Online.". I am running SharePoint with April 2020 update and I have the same behavior as you had.
> 
> The positive side is that [AppTenantTheme](https://docs.microsoft.com/en-us/sharepoint/dev/declarative-customization/site-theming/sharepoint-site-theming-rest-api#addtenanttheme) api endpoint is working, and I am able ti apply my custom made theme using [this](https://developer.microsoft.com/en-us/fluentui#/styles/themegenerator?fabricVer=5) tool.
> 
> There are two nuances with that API. First you'll need to call the "/_api/contextinfo" and to get the FormDigest and second you need to stringify the JSON ([example tool](https://onlinetexttools.com/json-stringify-text)) and POST it inside this payload:
> `{ "name":"feeee", "themeJson":"{\"palette\" : {\"themePrimary\": \"#e30613\",\"themeLighterAlt\": \"#fef4f5\",\"themeLighter\": \"#fbd4d6\",\"themeLight\": \"#f7afb4\",\"themeTertiary\": \"#ee646d\",\"themeSecondary\": \"#e6222f\",\"themeDarkAlt\": \"#cc0613\",\"themeDark\": \"#ac0510\",\"themeDarker\": \"#7f040c\",\"neutralLighterAlt\": \"#f8f8f8\",\"neutralLighter\": \"#f4f4f4\",\"neutralLight\": \"#eaeaea\",\"neutralQuaternaryAlt\": \"#dadada\",\"neutralQuaternary\": \"#d0d0d0\",\"neutralTertiaryAlt\": \"#c8c8c8\",\"neutralTertiary\": \"#dfb0f8\",\"neutralSecondary\": \"#c065f1\",\"neutralPrimaryAlt\": \"#a522eb\",\"neutralPrimary\": \"#9a07e8\",\"neutralDark\": \"#7405b0\",\"black\": \"#560482\",\"white\": \"#ffffff\",\"primaryBackground\": \"#ffffff\",\"primaryText\": \"#9a07e8\",\"bodyBackground\": \"#ffffff\",\"bodyText\": \"#9a07e8\",\"disabledBackground\": \"#f4f4f4\",\"disabledText\": \"#c8c8c8\"}}" }`

I tried it that way but still receive "false" as response from the server. Does the custom Theme show up on premise in your environment?

**Comment by tigertop** [comment]
Hi,
I have tested this for SharePoint 2019 on-prem. The theme never shows up but the colors are being applied. But I found a better way to handle themeing. Try this webpart [js-theme-manager-2019](https://github.com/pnp/sp-dev-fx-webparts/tree/master/samples/js-theme-manager-2019) is also a SharePoint Online variant in the their repo. This one only applies the theme you provide and the SharePoint Online variant also allows you to save the theme an set it from Look and Feel menu.
